### ISSUE TITLE
function return type does not seems to be respected

### ISSUE THREAD

**Issue by wjzhou-ep** [pylint-dev/pylint#9213]
### Bug description

For some reason, when use fsspec, pylint somehow infer the function return type wrong
```python
"""test"""
import fsspec.asyn

async def s3_ls_async() -> list[str]:
    """async"""
    return ["test"]

def s3_ls() -> list[str]:
    """list a folder"""
    loop = fsspec.asyn.get_loop()
    return fsspec.asyn.sync(loop, s3_ls_async)

def test_ls():
    """test"""
    data = s3_ls()
    assert data[0] == "test" # pylint error here "Value 'data' is unsubscriptable"
```

The s3_ls has a return type annotation of `list[str]`, but somehow pylint do not use it.

If the definition of s3_ls had been changed to the following, pylint does not report the error. FYI, I also, tried `typing.cast(list[str], ...)`, the same error was emiited

```
def s3_ls() -> list[str]:
    """direct return"""
    return ["test"]
```
```


### Configuration

_No response_

### Command used

```shell
pylint a_test.py
```

### Pylint output

```python
************* Module a_test
a_test.py:19:11: E1136: Value 'data' is unsubscriptable (unsubscriptable-object)
```

### Expected behavior

no error reported

### Pylint version

```shell
Tried:
pylint 3.0.2
astroid 3.0.1
Python 3.10.10 | packaged by conda-forge | (main, Mar 24 2023, 20:00:38) [MSC v.1934 64 bit (AMD64)]

ylint 2.17.4
astroid 2.15.6
Python 3.10.10 | packaged by conda-forge | (main, Mar 24 2023, 20:08:06) [GCC 11.3.0]
```

### OS / Environment

Windows/Linux

### Additional dependencies

```python
fsspec==2023.6.0
```

**Comment by wjzhou-ep** [comment]
@Pierre-Sassoulas Do you mind to give the a link to the duplicated issue?

**Comment by Pierre-Sassoulas** [comment]
Yeah, sorry I was on mobile. Duplicate of https://github.com/pylint-dev/pylint/issues/4813, pylint do not take typing into account and is not a type checker.
